Added at_train_end logic to base pipeline

maturk opened the pull request:
Here I have added at_train_end() logic to the base pipeline. To accommodate various different user needs, I added similar methods to the base_datamanager and base_model which are callable with **kwargs. 

Here is a small example of how I am personally using it within my custom pipeline for color correction:

```@profiler.time_function
    def at_train_end(self) -> None:
        self.eval()
        camera_ray_bundle, batch = self.datamanager.at_train_end()
        preds, refs = self.model.at_train_end(camera_ray_bundle=camera_ray_bundle, batch=batch)
        cc_images = self.color_correct(img=preds.cpu(), ref=refs.cpu())
        save_image((cc_images *255).astype(np.uint8), self.save_path, log=True)

tancik commented on nerfstudio/data/datamanagers/base_datamanager.py:
```diff
@@ -286,6 +286,9 @@ def get_param_groups(self) -> Dict[str, List[Parameter]]:  # pylint: disable=no-
         """
         return {}
 
+    def at_train_end(self, **kwargs: Any) -> Optional[Any]:  # pylint: disable=unused-argument disable=no-self-use
+        """Called at end of training for optional datamanager outputs."""
+
```
This is never called, I worry that users will override this function and expect it to be called.

tancik commented on nerfstudio/models/base_model.py:
```diff
@@ -219,3 +219,6 @@ def update_to_step(self, step: int) -> None:
         Args:
             step: training step of the loaded checkpoint
         """
+
+    def at_train_end(self, **kwargs: Any) -> Optional[Any]:  # pylint: disable=unused-argument disable=no-self-use
+        """Called at end of training for optional model outputs."""
```
Also never called, same potential confusion as above.

tancik reviewed (commented):
We should either make sure that the datamanager and model `train_end` are called, or remove them and leave it to the user to add such a function and call it in the pipeline.

maturk commented:
Thanks for the feedback again. I have added a new callback type OnTrainEndCallback. I removed unnecessary on_train_end calls from base model and base pipeline. This callback has access to the pipeline object.

Here is an example of how I am using the new callback type within my custom pipeline similar to my previous implementation:

````
    def get_on_train_end_callbacks(self) -> List[OnTrainEndCallback]:
        on_train_end_callbacks = []
        def color_correct_images():
            self.eval()
            camera_ray_bundle, batch = self.datamanager.at_train_end() # No longer in base datamanager, just a user defined function
            preds, refs = self.model.at_train_end(camera_ray_bundle=camera_ray_bundle, batch=batch) # No longer in base model, just a user defined function
            preds = self.color_correct(img=preds.cpu(), ref=refs.cpu())
            save_image((preds *255).astype(np.uint8), self.cc_save_paths, log=True)
        
        on_train_end_callbacks.append(
            OnTrainEndCallback(func=color_correct_images)
        )
        return on_train_end_callbacks

jkulhanek commented:
I am very sorry I didn't explain it clearly. I meant the already existing callback infrastructure. ...registering the callback here: https://github.com/nerfstudio-project/nerfstudio/blob/8b9574bb5a33a21bd1efae9bd777328ee566c0a7/nerfstudio/engine/callbacks.py#L42 , etc...

jkulhanek commented:
I think that would integrate in a more concise way with the rest of the code.

maturk commented:
Whew, no problem :) Happy to learn, maybe third time is the the charm! So I integrated it into the existing callback locations. Only thing I am not sure about is regarding the "step" var when calling run_callback_at_location(step,location) at the end of training. There is really no way to know if step==max-train-iterations inside the TrainingCallback class since this is not exposed. I have set step=None for now to signify that this is being called at train end.

jkulhanek commented on nerfstudio/engine/callbacks.py:
```diff
@@ -44,6 +44,7 @@ class TrainingCallbackLocation(Enum):
 
     BEFORE_TRAIN_ITERATION = auto()
     AFTER_TRAIN_ITERATION = auto()
+    ON_TRAIN_END = auto()
```
Can we name it AFTER_TRAIN_END?

jkulhanek replied:
Or perhaps "AFTER _TRAIN"?

jkulhanek commented on nerfstudio/engine/trainer.py:
```diff
@@ -297,6 +297,10 @@ def train(self) -> None:
 
+        # on train end callbacks
+        for callback in self.callbacks:
+            callback.run_callback_at_location(step=None, location=TrainingCallbackLocation.ON_TRAIN_END)
```
Can you pass the actual step here?

jkulhanek commented on nerfstudio/engine/callbacks.py:
```diff
@@ -91,8 +92,10 @@ def run_callback(self, step: int) -> None:
+            self.func(*self.args, **self.kwargs, step=step)
 
-    def run_callback_at_location(self, step: int, location: TrainingCallbackLocation) -> None:
+    def run_callback_at_location(self, step: Union[int, None], location: TrainingCallbackLocation) -> None:
```
Please keep the signature as int

jkulhanek commented on nerfstudio/engine/callbacks.py:
```diff
@@ -71,15 +72,15 @@ def __init__(
         self.kwargs = kwargs if kwargs is not None else {}
 
-    def run_callback(self, step: int) -> None:
+    def run_callback(self, step: Union[int, None]) -> None:
```
Please keep int here.

jkulhanek commented on nerfstudio/engine/callbacks.py:
```diff
@@ -71,15 +72,15 @@ def __init__(
         assert (
             "step" in signature(func).parameters.keys()
-        ), f"'step: int' must be an argument in the callback function 'func': {func.__name__}"
+        ), f"'step: Union[int, None]' must be an argument in the callback function 'func': {func.__name__}"
```
Same here, please keep int

tancik reviewed (approved):
LGTM
